url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/parser/Eval2.java

```diff
@@ -0,0 +1,1640 @@
+        }
+    }
+}
+
+class TimerHandler
+implements Eval2.IFrameHandler
+{
+    public void init(Eval2.EvalStack aStack, Eval2.StackFrame aFrame)
+    throws CommandException
+    {
+        final List lLstExpr = (List) aFrame.getExpr();
+        final int lLstSize = lLstExpr.size();
+
+        if(lLstSize != 2) throw new CommandException("The 'timer' form should have the format (timer expr).", aStack);
+
+        // Allocate a single slot for the expression we have to time and another slot for the start time.
+        aFrame.allocateData(2);
+    }
+
+    public void handleFrame(IEval aEval, Eval2.EvalStack aStack, Eval2.StackFrame aFrame)
+    throws CommandException
+    {
+        final List lLstExpr = (List) aFrame.getExpr();
+        final IContext lCtx = aFrame.getCtx();
+        final Object[] lData = aFrame.getData();
+        final int lDataPtr = aFrame.getDataptr();
+
+        if(lDataPtr == 0)
+        {
+            long lStart = System.currentTimeMillis();
+            aFrame.pushData(lStart);
+            aStack.push(lLstExpr.get(1), lCtx);
+            return;
+        }
+        else
+        {
+            long lStop = System.currentTimeMillis();
+            long lStart = (Long) lData[0];
+            aFrame.setResult(lStop - lStart);
+        }
+    }
+}
+
+class MacroHandler
+implements Eval2.IFrameHandler
+{
+    private CommandRepository macros = new CommandRepository();
+
+    public MacroHandler()
+    {
+    }
+
+    public CommandRepository getMacros()
+    {
+        return macros;
+    }
+
+    public void setMacros(CommandRepository aMacros)
+    {
+        macros = aMacros;
+    }
+
+    public boolean hasMacro(String aName)
+    {
+        return macros.hasCommand(aName);
+    }
+
+    public void init(Eval2.EvalStack aStack, Eval2.StackFrame aFrame)
+    throws CommandException
+    {
+        // Allocate a single slot for the evaluation of the macro expansion.
+        aFrame.allocateData(1);
+    }
+
+    @SuppressWarnings("unchecked")
+    public void handleFrame(IEval aEval, Eval2.EvalStack aStack, Eval2.StackFrame aFrame)
+    throws CommandException
+    {
+        final List lLstExpr = (List) aFrame.getExpr();
+        final IContext lCtx = aFrame.getCtx();
+        final Object[] lData = aFrame.getData();
+        final int lDataPtr = aFrame.getDataptr();
+
+        if(lDataPtr == 0)
+        {
+            try
+            {
+                // Built-in macro call.
+                final ICommand lMacro = macros.getCommand((String)lLstExpr.get(0));
+                final List lArgs = new ArrayList(lLstExpr.size());
+                lArgs.addAll(lLstExpr);
+                aStack.push(lMacro.execute(aEval, lCtx, lArgs.toArray()), lCtx);
+            }
+            catch(CommandException e)
+            {
+                Eval2.EvalStack lNestedStack = e.getStack();
+                e.setStack(aStack.merge(lNestedStack));
+                throw e;
+            }
+            catch(Exception e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)